pkg/operation/botanist/etcdencryption.go

```diff
@@ -36,85 +38,197 @@ import (
+		} else {
+			b.Logger.Infof("Encryption key for etcd secrets is already present in ShootState %s/%s", shootState.Namespace, shootState.Name)
+		}
+
+		if secretsEncryptionState == nil {
+			isEncryptionActive := metav1.HasAnnotation(secret.ObjectMeta, common.EtcdEncryptionChecksumAnnotationName)
+			forcePlaintextSecrets := kutil.HasMetaDataAnnotation(secret, common.EtcdEncryptionForcePlaintextAnnotationName, "true")
+
+			secretsEncryptionStatus := &encryptionconfiguration.ETCDEncryptionState{
+				ForcePlainTextResources: forcePlaintextSecrets,
+				IsEncryptionActive:      isEncryptionActive,
+			}
+
+			infodata.UpsertInfoData(&gardenerResourceDataList, common.ETCDSecretsEncryptionStateName, secretsEncryptionStatus)
+		} else {
+			b.Logger.Infof("Encryption state for etcd secrets is already present in ShootState %s/%s", shootState.Namespace, shootState.Name)
+		}
+
+		shootState.Spec.Gardener = gardenerResourceDataList
+		return nil
+	})
 	if err != nil {
 		return err
 	}
-
-	return b.syncEncryptionConfigurationToGarden(ctx, secret)
+	b.ShootState = shootState
+	return nil
 }
 
-func (b *Botanist) syncEncryptionConfigurationFromGardenCluster(ctx context.Context) error {
-	gardenSecret := &corev1.Secret{}
-	if err := b.K8sGardenClient.Client().Get(ctx, common.GardenEtcdEncryptionSecretKey(b.Shoot.Info.Namespace, b.Shoot.Info.Name), gardenSecret); err != nil {
+// SyncForcePlaintextSecrets syncs the annotation which determines whether secret resources should be kept unencrypted in the
+// Shoot's ETCD from the ETCD encryption configuration secret.
+func (b *Botanist) SyncForcePlaintextSecrets(ctx context.Context) error {
+	secret := &corev1.Secret{ObjectMeta: kutil.ObjectMeta(b.Shoot.SeedNamespace, common.EtcdEncryptionSecretName)}
+	err := b.K8sSeedClient.Client().Get(ctx, kutil.KeyFromObject(secret), secret)
+	if err != nil {
 		if apierrors.IsNotFound(err) {
 			return nil
 		}
 		return err
 	}
+	shootState := &gardencorev1alpha1.ShootState{ObjectMeta: kutil.ObjectMeta(b.Shoot.Info.Namespace, b.Shoot.Info.Name)}
+	_, err = controllerutil.CreateOrUpdate(ctx, b.K8sGardenClient.Client(), shootState, func() error {
+		gardenerResourceDataList := gardencorev1alpha1helper.GardenerResourceDataList(shootState.Spec.Gardener)
+		forcePlaintextSecrets := kutil.HasMetaDataAnnotation(secret, common.EtcdEncryptionForcePlaintextAnnotationName, "true")
+		encryptionStateInfoData, err := infodata.GetInfoData(gardenerResourceDataList, common.ETCDSecretsEncryptionStateName)
+		if err != nil {
+			return err
+		}
+		etcdEncryptionState, ok := encryptionStateInfoData.(*encryptionconfiguration.ETCDEncryptionState)
+		if !ok {
+			return fmt.Errorf("Could not convert GardenerResourceData entry %s to encryptionconfiguration.ETCDEncryptionState", common.ETCDSecretsEncryptionStateName)
+		}
+		etcdEncryptionState.ForcePlainTextResources = forcePlaintextSecrets
```

Comment:
I assume that instead of `forcePlaintextSecrets` you mean `isEncryptionActive` as that is what you linked.
> in the migration code when there is no etcdEncryptionState (the marked code) -> calculate from the etcd secret

This will be removed in the future.

Here's the way that the ETCD encryption secret was generated previously (before the shootstate)
On first reconciliation (creation)
1. Generate a passive encryption configuration and generate a secret object from it.
2. Calculate the checksum of the encryption configuration and only save it in memory
3. Rewrite the shoot secrets and [add the checksum annotation to the secret object](https://github.com/gardener/gardener/blob/753f00bbd5a0745361022aa698faf701c63c3971/pkg/operation/boanist/etcdencryption.go#L336).

On second reconciliation:
1. See that the secret object has a checksum annotation
2. Change the passive encryption configuration to active, update the secret
3. Recalculate the checksum and save it in memory
4. Rewrite shoot secrets with enabled encryption and [update the checksum as it has changed](https://github.com/gardener/gardener/blob/753f00bbd5a0745361022aa698faf701c63c3971/pkg/operation/botanist/etcdencryption.go#L336).

I've tried to keep the same workflow as I'm not sure if there will be any side effects if I change it slightly.
But indeed, I should not update `isEncryptionActive` every time. I used to update the checksum to the etcd encryption secret in the garden cluster every time but this isn't needed here. Thanks for pointing it out :).

Reply:
Thanks for the detailed explanation!